### PR TITLE
UPSTREAM: 52917: add --selector support to drain,cordon

### PR DESCRIPTION
**Pull request by juanvallejo**
Related Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1466390
Related Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1466382

Adds `--selector` option to `oc adm drain,cordon,uncordon`, performing
each action on a list of nodes matching the given label.

If at least one node fails to be cordoned/uncordoned, the command will
continue to operate on any remaining nodes, uninterrupted, reporting
any errors it encounters along the way.

If at least one node fails to be drained, the command will halt, printing
the immediate error encountered, and a list of nodes yet to be drained
(including the node that just failed to be drained).

Work in progress.
TODO

- [x] add tests

cc @openshift/cli-review @soltysh

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go [hunk 1]
```diff
@@ -208,27 +222,47 @@ func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
...
-	r := o.Factory.NewBuilder(true).
+	nameArgs := []string{"nodes"}
+
+	b := o.Factory.NewBuilder(true).
 		NamespaceParam(cmdNamespace).DefaultNamespace().
-		ResourceNames("node", args[0]).
-		Do()
+		SelectorParam(o.Selector)
+
+	if len(args) > 0 {
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: Move that if before you create the builder and you can then squash the whole builder invocation into one.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go [hunk 2]
```diff
@@ -208,27 +222,47 @@ func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
...
 		return err
 	}
 
+	if len(o.Selector) > 0 {
+		infos, err := r.Infos()
+		if err != nil {
+			return err
+		}
+		if len(infos) == 0 {
+			return fmt.Errorf("error: no nodes matched selector %q", o.Selector)
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: Do you think we need that message? I haven't seen this being error-ed out anywhere else when applying label selector.

Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: Thanks, will remove

**Comment by juanvallejo** [comment]
@soltysh thanks for the review, will open upstream PR

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go [hunk 3]
```diff
@@ -188,15 +193,24 @@ func NewCmdDrain(f cmdutil.Factory, out, errOut io.Writer) *cobra.Command {
...
 	return cmd
 }
 
 // SetupDrain populates some fields from the factory, grabs command line
 // arguments and looks up the node using Builder
 func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
+	o.Selector = cmdutil.GetFlagString(cmd, "selector")
+
 	var err error
-	if len(args) != 1 {
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: With this new if you're swallowing the case where nothing was passed and you should return error. I think the best option would be to:

1. check args == 0 -> error out
2. check if args and selector is passed -> error out

File: test/cmd/admin.sh [hunk 4]
```diff
@@ -84,6 +85,18 @@ echo "manage-node: ok"
 os::test::junit::declare_suite_end
 
 os::test::junit::declare_suite_start "cmd/admin/certs"
+# check node-management cmds: drain,cordon,uncordon
+# ensure drain fails when both a nodeName and --selector are given
+os::cmd::expect_failure_and_text 'oc adm drain fake-node --selector fake=label' 'cannot specify both a node name and a \-\-selector option'
```
Comment (soltysh) [inline test/cmd/admin.sh]: Can you also add the other failure, when no parameters are passed?

**Comment by soltysh** [comment]
I'm tagging this as do-not-merge, until an actual upstream commit is present. I don't want to end up with this being merged before upstream will be created and approved.

**Comment by soltysh** [comment]
Still waiting for upstream approval.

**Comment by juanvallejo** [comment]
@liggitt @soltysh is this good to go?

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go [hunk 5]
```diff
@@ -208,18 +222,28 @@ func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
...
+		nameArgs = append(nameArgs, args[0])
+		if strings.Contains(args[0], "/") {
+			nameArgs = []string{args[0]}
+		}
+	}
+
 	r := o.Factory.NewBuilder(true).
 		NamespaceParam(cmdNamespace).DefaultNamespace().
-		ResourceNames("node", args[0]).
+		SelectorParam(o.Selector).
```
Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: shoudn't this be `ResourceNames("nodes", args...)`?

Comment (liggitt) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: the `nameArgs := []string{"nodes"}` and `if strings.Contains(args[0], "/")` block above means this command parses resource args differently than other commands... 

I'd expect `ResourceNames("nodes", args...).SingleResourceType()` and a check on the resource type of selected resources to make sure they are nodes

**Comment by juanvallejo** [comment]
@liggitt thanks, review comments addressed. Will go ahead and update drain.go upstream if everything looks okay.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/drain.go#L240-L253

**Comment by juanvallejo** [comment]
ping @liggitt @soltysh

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go (hunk 5 above)
Comment (liggitt) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: link to the upstream PR that resolved this? also, did 52917 go into 1.8?

Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/drain.go]: Upstream patch: https://github.com/kubernetes/kubernetes/pull/54083
52917 is not in 1.8: [https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/kubectl/cmd/drain.go]()

**Comment by juanvallejo** [comment]
Closing as the [current rebase](https://github.com/openshift/origin/pull/17576) will be introducing these commits to origin

**Pull request closed without merging**